src/Foundatio/Utility/SharedOptions.cs

```diff
@@ -0,0 +1,28 @@
+using System;
+using Foundatio.Serializer;
+using Microsoft.Extensions.Logging;
+
+namespace Foundatio {
+    public class SharedOptions {
+        public ISerializer Serializer { get; set; }
+        public ILoggerFactory LoggerFactory { get; set; }
+    }
+
+    public interface ISharedOptionsBuilder : IOptionsBuilder {}
```

Comment:
I am think about whether the `ISharedOptionsBuilder` and `SharedOptionsBuilderExtensions` can be just combined into single class `SharedOptionsBuilder` to reduce the complexity of inherit. I am worrying about it is too complex to create new module and difficult to understand for new comers. For example there should have to 4 classes/interfaces included `ISharedQueueOptions`, `SharedQueueOptions<T>`, `ISharedQueueOptionsBuilder`, `SharedQueueOptionsBuilderExtensions`.

Reply:
Yeah, I agree with that. I think it would be better if we just had base options classes with methods and not use the extension methods.

Reply:
I just tried to combine into a single options class that you would just derive from, but it doesn't really work because without the methods being extension methods, you can't have them reflect back the same type that they were invoked with.

Reply:
The actual options classes end up being very simple:

```csharp
public class FolderFileStorageOptionsBuilder : OptionsBuilder<FolderFileStorageOptions>, ISharedOptionsBuilder {
    public FolderFileStorageOptionsBuilder Folder(string folder) {
        if (string.IsNullOrEmpty(folder))
            throw new ArgumentNullException(nameof(folder));
        Target.Folder = folder;
        return this;
    }
}
```

You can bring in the shared options by simply adding the marker interface to your builder and deriving your options class from the sharedoptions. The concrete methods on your builder class end up being very simple too.

Reply:
Excellent. So the other modules should be improved like this(Queues, Metrics and Messaging)

Reply:
They should already be the same pattern, are they not? Queues are a bit more complicated since they have generics with the model type.

Reply:
I think the `ISharedQueueOptionsBuilder`, `SharedQueueOptions<T>`, `ISharedQueueOptionsBuilder` and `SharedQueueOptionsBuilderExtensions` can be combined to the following code
```cs
    public class SharedQueueOptions<T> : SharedOptions where T : class {
        public string Name { get; set; } = typeof(T).Name;
        public int Retries { get; set; } = 2;
        public TimeSpan WorkItemTimeout { get; set; } = TimeSpan.FromMinutes(5);
        public IEnumerable<IQueueBehavior<T>> Behaviors { get; set; }
    }

    public class SharedQueueOptionsBuilder<T, TOptions, TBuilder> : OptionsBuilder<TOptions> 
        where T : class
        where TOptions : SharedQueueOptions<T>, new()
        where TBuilder : SharedQueueOptionsBuilder<T, TOptions, TBuilder> {
        public TBuilder Name(string name) {
            Target.Name = name;
            return (TBuilder)this;
        }

        public TBuilder Retries(int retries) {
            Target.Retries = retries;
            return (TBuilder)this;
        }

        public TBuilder WorkItemTimeout(TimeSpan timeout) {
            Target.WorkItemTimeout = timeout;
            return (TBuilder)this;
        }

        public TBuilder Behaviors(IEnumerable<IQueueBehavior<T>> behaviors) {
            Target.Behaviors = behaviors;
            return (TBuilder)this;
        }

        public TBuilder AddBehavior(IQueueBehavior<T> behavior) {
            if (behavior == null)
                throw new ArgumentNullException(nameof(behavior));
            if (Target.Behaviors == null)
                Target.Behaviors = new[] { behavior };
            else
                Target.Behaviors = Target.Behaviors.Concat(new[] { behavior });

            return (TBuilder)this;
        }
    }
```
And then the fluent methods for `InMemoryQueueOptions` should be simplified like
```cs
    public class InMemoryQueueOptionsBuilder<T> : SharedQueueOptionsBuilder<T, InMemoryQueueOptions<T>, InMemoryQueueOptionsBuilder<T>> where T: class {
        public InMemoryQueueOptionsBuilder<T> RetryDelay(TimeSpan retryDelay) {
            Target.RetryDelay = retryDelay;
            return this;
        }

        public InMemoryQueueOptionsBuilder<T> RetryMultipliers(int[] multipliers) {
            Target.RetryMultipliers = multipliers ?? throw new ArgumentNullException(nameof(multipliers));
            return this;
        }
    }
```

Reply:
The `ISharedOptionsBuilder` and `SharedOptionsBuilderExtensions` should be combined to the following code at first
```cs
    public class SharedOptionsBuilder<TOption, TBuilder> : OptionsBuilder<TOption>
        where TOption : SharedOptions, new()
        where TBuilder : SharedOptionsBuilder<TOption, TBuilder> {
        public TBuilder Serializer(ISerializer serializer) {
            Target.Serializer = serializer ?? throw new ArgumentNullException(nameof(serializer)); ;
            return (TBuilder)this;
        }

        public TBuilder LoggerFactory(ILoggerFactory loggerFactory) {
            Target.LoggerFactory = loggerFactory ?? throw new ArgumentNullException(nameof(loggerFactory)); ;
            return (TBuilder)this;
        }
    }
```
And the `SharedQueueOptionsBuilder` should be changed to
```cs
    public class SharedQueueOptionsBuilder<T, TOptions, TBuilder> : SharedOptionsBuilder<TOptions, TBuilder> 
        where T : class
        where TOptions : SharedQueueOptions<T>, new()
        where TBuilder : SharedQueueOptionsBuilder<T, TOptions, TBuilder> {
    }
```

Reply:
Please refer to the commit for the full list changes
https://github.com/edwardmeng/Foundatio/commit/27b6196b19fbb8a40bd522c96fb0331d5e2d9673

Reply:
Yeah, my issue with that is that it forces inheritance from SharedOptions and SharedOptionsBuilder vs right now you can just add the marker interface to your builder and you could still derive from something else. Same thing for the options classes, you could just implement the ISharedOptions (just added) interface and not have to derive from SharedOptions and the builder methods would still work.

The other problem I have is that its making the concrete builder and options classes a lot more verbose with them having to specify both the builder and options types in the inherited classes.

The queue options is really the only one that is kind of breaking things because it uses generics. I tried to figure out a way for it to not use generics, but it seems like it needs them for the behaviors. If you look at the concrete builders and options classes, they are very simple. It's just the shared stuff that is a little verbose / ugly, but that is what we are supposed to be doing. We do the hard stuff and try and make the implementations as straight forward as possible.

Also, I just simplified the behaviors a bit by reducing it down to a single builder method that can take either a single behavior, multiple or an array of them by using a `params` array.